bindgen/src/gen_go/mod.rs

```diff
@@ -38,30 +69,113 @@ pub struct CustomTypeConfig {
+fn load_toml_file(source: &Utf8Path) -> Result<toml::value::Table> {
+    let contents = fs::read_to_string(source)
+        .with_context(|| format!("Failed to read config file: {:?}", source))?;
+    let mut cfg: toml::value::Table = toml::de::from_str(&contents)?;
+
+    let res = cfg
+        .remove("bindings")
+        .and_then(|mut cfg| cfg.as_table_mut().map(|tbl| tbl.remove(Config::TOML_KEY)))
+        .flatten()
+        .and_then(|cfg| cfg.try_into().ok())
+        .unwrap_or_else(|| toml::value::Table::default());
+    Ok(res)
+}
+
 impl Config {
+    pub fn load_initial(
+        crate_root: &Utf8Path,
+        config_file_override: Option<&Utf8Path>,
+    ) -> Result<Self> {
+        let default_config = crate_root.join("uniffi.toml").canonicalize_utf8().ok();
+        let toml_config = match (default_config, config_file_override) {
+            (Some(cfg), Some(over)) => {
+                let mut cfg: toml::value::Table = load_toml_file(&cfg)?;
+                let over: toml::value::Table = load_toml_file(over)?;
+
+                fn merge(a: &mut toml::value::Table, b: toml::value::Table) {
+                    for (key, value) in b.into_iter() {
+                        match a.get_mut(&key) {
+                            Some(existing_value) => match (existing_value, value) {
+                                (toml::Value::Table(ref mut t0), toml::Value::Table(t1)) => {
+                                    merge(t0, t1);
+                                }
+                                (v, value) => *v = value,
+                            },
+                            None => {
+                                a.insert(key, value);
+                            }
+                        }
+                    }
+                }
+
+                merge(&mut cfg, over);
+
+                toml::Value::from(cfg)
+            }
+            (Some(cfg), None) => toml::Value::from(load_toml_file(&cfg)?),
+            (None, Some(over)) => toml::Value::from(load_toml_file(&over)?),
+            (None, None) => toml::Value::from(toml::value::Table::default()),
+        };
+
+        Ok(toml_config.try_into()?)
+    }
+
+    /// The name of the go package containing the high-level foreign-language bindings.
     pub fn package_name(&self) -> String {
-        if let Some(package_name) = &self.package_name {
-            package_name.clone()
-        } else {
-            "uniffi".into()
+        match self.module_name.as_ref() {
+            Some(name) => name.clone(),
+            None => "uniffi".into(),
         }
     }
-}
 
-impl MergeWith for Config {
-    fn merge_with(&self, other: &Self) -> Self {
-        Config {
-            package_name: self.package_name.merge_with(&other.package_name),
-            custom_types: self.custom_types.merge_with(&other.custom_types),
+    /// The name of the lower-level C module containing the FFI declarations.
+    pub fn ffi_package_name(&self) -> String {
```

Comment:
should be all `package_name` now

Reply:
The configuration options still look very convoluted to me. `ffi_package_name` seems to be a default value for `ffi_module_filename`, its not used anywhere else and I think it can be removed.

I recently added user focused documentation for these configuration options, and it really clears things up. I suggest to do the same for these new options.
https://github.com/nordSecurity/uniffi-bindgen-cs#configuration-options